users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py

```diff
@@ -0,0 +1,78 @@
+class PerturbationFactor:
+  """
+  Class to wrap perturbation factors, e.g. for speed or tempo perturbation.
+  """
+  def __init__(self, prob, minimum, maximum):
+    self.prob = prob
+    self.min = minimum
+    self.max = maximum
+
+
+class WaveformPerturbation:
+  """
+  Helper class to perform perturbation techniques on audio waveforms.
+  """
+  def __init__(self, speed=None, tempo=None, sox_effects=None, codecs=None, preemphasis=None):
+    """
+    :param speed:
+    """
+    import torch
+    import functools
+    self._speed = PerturbationFactor(**speed) if speed else None
+    self._tempo = PerturbationFactor(**tempo) if tempo else None
+    self._perturbations = [functools.partial(self.sox, sox_effects=sox_effects)]
+    if preemphasis:
+      self._perturbations.append(functools.partial(self.preemphasis, factor=PerturbationFactor(**preemphasis)))
+    if codecs:
+      self._perturbations.append(functools.partial(self.apply_codecs, codecs=codecs))
+
+  def run(self, audio, sample_rate, random_state):
+    audio = torch.from_numpy(audio).unsqueeze(0).to(torch.float32)
+    for perturbation in self._perturbations:
+      audio = perturbation(audio, sample_rate, random_state)
+    audio = audio.numpy().squeeze()
+    assert isinstance(audio, np.ndarray)
+    assert len(audio.shape) == 1
+    return audio
+
+  def sox(self, audio, sample_rate, random_state, sox_effects):
+    import torchaudio
+    sox_effects = sox_effects or []
+    speed = False
+    if self._speed is not None:
+      if random_state.random() < self._speed.prob:
+        factor = random_state.random() * (self._speed.max - self._speed.min) + self._speed.min
+        sox_effects.append(["speed", str(factor)])
+        sox_effects.append(["rate", str(sample_rate)])
+        speed = True
+    if self._tempo is not None:
+      if random_state.random() < self._tempo.prob and not speed:
+        factor = random_state.random() * (self._tempo.max - self._tempo.min) + self._tempo.min
+        sox_effects.append(["tempo", str(factor)])
+    audio, _ = torchaudio.sox_effects.apply_effects_tensor(audio, sample_rate, sox_effects)
+    return audio
+
+  @staticmethod
+  def preemphasis(audio, sample_rate, random_state, factor):
+    import torch
```

Comment:
I guess it should be `torchaudio` here? In general, I think it would be a bit nicer not to have the imports in each function but once at the beginning. You could achieve this by adding a string with the imports as the first entry in the list returned by `get_classes_perturbation()` (which might then be renamed to better reflect what it does).

Reply:
Changed it so the imports are added separately